hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/CustomizePathsTraverser.java

```diff
@@ -0,0 +1,287 @@
+                        }
+                        adjacency.add(newNode);
+
+                        // Avoid exceeding capacity
+                        if (capacity != NO_LIMIT &&
+                            ++accessCount >= capacity) {
+                            break root;
+                        }
+                    }
+                }
+
+                if (step.sample > 0) {
+                    // Sample current node's adjacent nodes
+                    adjacency = sample(adjacency, step.sample);
+                }
+
+                // Add current node's adjacent nodes
+                for (Node node : adjacency) {
+                    newVertices.add(node.id(), node);
+                    // Avoid exceeding limit
+                    if (stepNum == 0) {
+                        if (limit != NO_LIMIT && !sorted &&
+                            ++pathCount >= limit) {
+                            break root;
+                        }
+                    }
+                }
+            }
+            // Re-init sources
+            sources = newVertices;
+        }
+        if (stepNum != 0) {
+            return ImmutableList.of();
+        }
+        List<Path> paths = new ArrayList<>();
+        for (List<Node> nodes : newVertices.values()) {
+            for (Node n : nodes) {
+                if (sorted) {
+                    WeightNode wn = (WeightNode) n;
+                    paths.add(new WeightPath(null, wn.path(), wn.weights()));
+                } else {
+                    paths.add(new Path(null, n.path()));
+                }
+            }
+        }
+        return paths;
+    }
+
+    public static List<Path> topNPath(List<Path> paths,
+                                      boolean incr, long limit) {
+        if (limit == HugeTraverser.NO_LIMIT || paths.size() <= limit) {
+            return paths;
+        }
+
+        paths.sort(((p1, p2) -> {
+            CustomizePathsTraverser.WeightPath wp1 = (CustomizePathsTraverser.WeightPath) p1;
+            CustomizePathsTraverser.WeightPath wp2 = (CustomizePathsTraverser.WeightPath) p2;
+            int result;
+            if (wp1.totalWeight() > wp2.totalWeight()) {
+                result = 1;
+            } else if (wp1.totalWeight() < wp2.totalWeight()) {
+                result = -1;
+            } else {
+                result = 0;
+            }
+            if (incr) {
+                return result;
+            } else {
+                return -result;
+            }
+        }));
+
+        return paths.subList(0, (int) limit);
+    }
+
+    private static List<Node> sample(List<Node> nodes, long sample) {
+        if (nodes.size() <= sample) {
+            return nodes;
+        }
+        List<Node> result = new ArrayList<>((int) sample);
+        for (int random : randomSet(0, nodes.size(), (int) sample)) {
+            result.add(nodes.get(random));
+        }
+        return result;
+    }
+
+    // TODO: move to common
```

Comment:
moved to https://github.com/hugegraph/hugegraph-common/pull/16/commits/55d50be090203594ead23d88a0db631bcae67f9b#diff-30e4547f56488c72b2609834615deaacR72